### ISSUE TITLE
Is the insteon hub supported?

### ISSUE THREAD

**Issue by thedannymullen** [TD22057/insteon-mqtt#77]
Is the insteon hub supported?  The 2014+ model 2245-222?  If not any chance to add it?

**Comment by TD22057** [comment]
Sorry - no (at least not by me).  The newer hub doesn't expose a PLM API and so no real time updates are possible as far as I know.  This lib is designed around having something that really works and, in my opinion, a polling API with the hub doesn't work well at all.  Feel free to work on some way to integrate it if you'd like to - I don't mind merging something if it's well done.   The whole MQTT system is modular - there are no dependencies on the PLM system.  So you could write something that polls the hub and uses everything else from this lib to handle the MQTT integration.  I'm not sure if the hub gives access to the low level insteon link tables or not.  My recommendation is to spend the money and get a PLM modem.  

I'll leave this open in case anyone else wants to take a cut at it.  But if your time is worth anything, it'd be cheaper to just by a PLM modem.

**Comment by thedannymullen** [comment]
Thanks for responding to the issue!  I don’t have time to write the code at the present moment.  

If possible that would be great to leave this open.  I appreciate that offer.  

I will research plm modems and see what they run.

**Comment by lnr0626** [comment]
According to https://www.smarthome.com/powerlinc-modem-insteon-2413u-usb-interface-dual-band.html the plm is dual band (i.e. supports both wireless and power line)

**Comment by bfranske** [comment]
FYI. I did a little research on this unrelated to this project but it may prove useful. Officially there is no local API for the Insteon hub, there is an official REST API which uses the Insteon servers and requires an API key. However, some people have done some work decoding ways to make HTTP requests to the local hub address and send commands. See:
- http://www.richstevenson.com/2014/01/06/insteon-direct-commands/
- http://leftovercode.info/smartlinc.php
- http://www.smarthome.com.au/smarthome-blog/cat/insteon/post/insteon-hub-http-commands/

[There was apparently an update in July 2018 that broke that interface though.](https://community.openhab.org/t/insteon-update/46665/37) After some complaints by the OpenHab and Homeseer communities Insteon relented and reactivated the way they were doing things and promised to check future updates for compatibility so now I guess it's "unofficially supported".

**Comment by krkeegan** [comment]
It can be done, it isn't ideal.  I was working on another project that had solved this, but this project is better designed, so I stopped working on it.  If you want to try and port the code, the relevant stuff is here:

https://github.com/krkeegan/insteon-mngr/blob/master/insteon_mngr/hub.py

In short, you can view the data buffer on the hub from a webpage.  The buffer endlessly loops and there is no push notifications of changes in the data.  So, you have to constantly query the hub's html page.  It handles 3 queries a second without issue.  In an ideal world, you would put the hub querying code on it's own thread. Then using a Queue to transfer the read and write data back and forth between it.

I have sucked at contributing time to this project recently, so I am not committing to anything.  If you have the skills I think the code is all there for you to incorporate this.  Otherwise, I will hopefully get to it eventually.

**Comment by thedannymullen** [comment]
@krkeegan I was re-visiting this.   Thanks for Info!  

@bfranske  thanks!

**Comment by jtmoderate876** [comment]
My PLM died and I started looking for a way to talk to my HUB instead but after reading this I will just reluctantly reward Insteon and buy another PLM. It will be interesting trying to back up my existing LinkDBs and somehow edit/restore it for the new PLM.

**Comment by JonS9999** [comment]
If memory serves, PLMs can die because of an underrated capacitors in
them.  Doing a quick Google search, I came across this:


https://1projectaweek.com/blog/2017/6/2/repairing-a-failed-insteon-powerlinc-power-line-modem-plm-2413u


My PLM just started acting up last night, so I'm going to swap out the caps
to see if it improves anything.

Cheers.

**Comment by jtmoderate876** [comment]
Good point - it might be worth it for me to try that in order to avoid having to re-link a new PLM with new address to every Insteon.

**Comment by TD22057** [comment]
At least w/ the latest update you can back up your whole network which should make it easier to handle a dead modem in the future (not much help now of course).

**Comment by JonS9999** [comment]
Just be sure you (don't) solder like this person...

https://static.boredpanda.com/blog/wp-content/uploads/2016/03/stock-image-fail-soldering-iron-bob-byron-1.jpg

LOL!

**Comment by jtmoderate876** [comment]
> At least w/ the latest update you can back up your whole network which should make it easier to handle a dead modem in the future (not much help now of course).

Good to know - thx!

> Just be sure you (don't) solder like this person...

That isn't how you hold it? :)

**Comment by krkeegan** [comment]
OK, I got a version of Hub support "working."  I put that in quotes because it needs work before it can be usable.

@TD22057 the biggest issue is how to incorporate this into the "main loop."  The interface to the hub sadly uses http requests.  There is no socket interface that I can find anywhere.  So to read from the hub, we send http get requests to the same page looking for new data.  To send data, make a get call with attributes in the URL.

This is fine, I have written a small client that runs in its own thread and does all of this work in its own loop.  It passes the data back to the main thread using Queues.

I have it setup as a polling device right now, but the select call for the MQTT link blocks for 3 seconds every loop, unless there is data.  I suppose I could convert the use of Queues into a socket and re-broadcast the hub data on a TCP socket and then we could use select to bind and read from this socket.  Is that the best option?

**Comment by krkeegan** [comment]
If you want to look at the code it is below.  The HubClient is at the end of the network/Hub.py file.

https://github.com/krkeegan/insteon-mqtt/tree/Add_Hub_Support

**Comment by jtmoderate876** [comment]
TL;DR - failed at PLM capacitor replacement so bought another PLM.

Regarding restoring my USP PLM. I bought the capacitors but failed. My PLM is version 1.0 and it is slightly different (pretty hoaky soldering of coils, capacitors and trace wires in "mid air"). I did find other links that address this 1.0 version and I think I did it right but in the end - I bought another PLM (just hate rewarding insteon/smarthome for failures - I did this with a hub a year or two ago (before they come up with the "send it to us" hub program where you keep your programming)).

**Comment by krkeegan** [comment]
I pushed another commit to the branch.  With a few additions.

- The loop timeout is now only .5 a second when a hub is used.  This seems to work perfectly without any increase in CPU usage.
- Added two Hub Specific messages so that they are properly parsed.  One is totally unknown.  Another is a message that shows up when a device is unreachable.  I am not sure it gets us anything, but it now gets properly parsed.

I need to write the unit tests and the documentation and I should be able to make a pull request.  But it all works now if anyone wants to try it out.

**Comment by krkeegan** [comment]
Alright, the pull request is done.  Everything seems to work well.
